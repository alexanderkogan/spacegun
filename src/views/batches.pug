doctype html
html(lang="en")
  include includes/head.pug
body
    h1
        a(href="/")="Home"
        span=" ∞ Batches"
    

    each n in namespaces
        h3=name+" ∞ "+n.name
        table(class="u-full-width")
            thead
                tr
                    th="Name"
                    th="Current Image Url"
                    th="Cron"
                    th="Concurrency"
            tbody
                each batch in n.batches
                    tr
                        td=batch.name
                        td=batch.image.url
                        td=batch.schedule
                        td=batch.concurrencyPolicy
