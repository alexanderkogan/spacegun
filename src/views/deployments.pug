doctype html
html(lang="en")
  include includes/head.pug
body
    h1
        a(href="/")="Home"
        span=" ∞ Pods"
    

    each n in namespaces
        h3=name+" ∞ "+n.name
        table(class="u-full-width")
            thead
                tr
                    th="Name"
                    th="Current Image Url"
            tbody
                each deployment in n.deployments
                    tr
                        td=deployment.name
                        td=deployment.image.url
