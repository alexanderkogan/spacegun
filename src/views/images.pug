doctype html
html(lang="en")
  include includes/head.pug
body
    h1
        a(href="/")="Home"
        span=` ∞ Images ∞ ${name}`
    
    if focusedImage
        h3= `${focusedImage.name} Ξ ${focusedImage.tag}`
        p=focusedImage.url

    table(class="u-full-width")
        thead
            tr
                th="Name"
                th="Tag"
        tbody
            each image in images
                if !focusedImage || focusedImage.tag !== image.tag
                    tr
                        td=image.name
                        td
                            a(href=`/images/${encodeURIComponent(image.name)}/${encodeURIComponent(image.tag)}`)=image.tag
