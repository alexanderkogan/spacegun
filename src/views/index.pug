doctype html
html(lang="en")
  include includes/head.pug
body
    h1="Welcome to Spacegun "+version

    if errors.length > 0
        h2="Errors"
        ul
            for error in errors
                li=error

    if clusters
        h2="Clusters"
        table(class="u-full-width")
            thead
                tr
                    th="Name"
                    th="Known Namespaces"
            tbody
                each cluster in clusters
                    tr
                        td
                            a(href=`pods/${cluster.name}`)=cluster.name
                        td=cluster.namespaces

    if jobs
        h2="Jobs"
        table(class="u-full-width")
            thead
                tr
                    th="Name"
                    th="Last Run"
                    th="Next Run"
            tbody
                each job in jobs
                    tr
                        td=job.job.name
                        td=job.lastRun
                        td=job.nextRun

    if images
        h2="Images"
        each image in images
            div(class="u-pull-left")
                a(href=`images/${image}`, class="card")=image
        div(class="u-cf")

    h2="Configuration"
    pre
        code
            span.
                kube: #{config.kube}
                docker: #{config.docker}
                jobs: #{config.jobs}

            if config.slack
                span.
                    slack: #{config.slack}

            if config.namespaces
                span.
                    namespaces: [#{config.namespaces}]

            if config.git
                span.
                    git:
                        remote: #{config.git.remote}
                        cron: #{config.git.cron}

            span.
                server:
                    host: #{config.server.host}
                    port: #{config.server.port}
