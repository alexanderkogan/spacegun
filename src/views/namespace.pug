doctype html
html(lang="en")
  include includes/head.pug
body
    h1
        a(href="/")="Home"
        span=" ∞ Namespace ∞ " + namespaceName
    
    h3=clusterName+" ∞ "+namespaceName+" ∞ "+"Deployments"
    table(class="u-full-width")
        thead
            tr
                th="Name"
                th="Current Image Url"
        tbody
            each deployment in deployments
                tr
                    td=deployment.name
                    td=deployment.image.url

    h3=clusterName+" ∞ "+namespaceName+" ∞ "+"Batches"
    table(class="u-full-width")
        thead
            tr
                th="Name"
                th="Current Image Url"
                th="Cron"
                th="Concurrency"
        tbody
            each batch in batches
                tr
                    td=batch.name
                    td=batch.image.url
                    td=batch.schedule
                    td=batch.concurrencyPolicy

    h3=clusterName+" ∞ "+namespaceName+" ∞ "+"Pods"
    table(class="u-full-width")
        thead
            tr
                th="Name"
                th="State"
                th="Current Image Url"
                th="Restarts"
        tbody
            each pod in pods
                tr
                    td=pod.name
                    if pod.ready
                        td="Up"
                    else
                        td="Down"
                    td=pod.image.url
                    td=pod.restarts
