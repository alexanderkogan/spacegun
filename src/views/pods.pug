doctype html
html(lang="en")
  include includes/head.pug
body
    h1
        a(href="/")="Home"
        span=" ∞ Pods"
    

    each n in namespaces
        h3=name+" ∞ "+n.name
        table(class="u-full-width")
            thead
                tr
                    th="Name"
                    th="State"
                    th="Current Image Url"
                    th="Restarts"
            tbody
                each pod in n.pods
                    tr
                        td=pod.name
                        if pod.ready
                            td="Up"
                        else
                            td="Down"
                        td=pod.image.url
                        td=pod.restarts
