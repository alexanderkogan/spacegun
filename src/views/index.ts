import * as moment from "moment"
import { Resource } from "@/dispatcher/resource"
import { call } from "@/dispatcher"
import { clusters, namespaces, pods } from "@/cluster/ClusterModule"
import { jobs, schedules } from "@/jobs/JobsModule"
import { Config } from "@/config"
import { list, tags } from "@/images/ImageModule"

let config: Config | undefined
export function init(c: Config) {
    config = c
}

export class Module {

    @Resource({ path: "/" })
    public async index(): Promise<object> {
        const errors: string[] = []

        let clustersWithNamespaces = undefined
        try {
            const knownClusters: string[] = await call(clusters)()
            clustersWithNamespaces = []
            for (const cluster of knownClusters) {
                const knownNamespaces = await call(namespaces)({ cluster })
                clustersWithNamespaces.push({
                    name: cluster,
                    namespaces: knownNamespaces
                })
            }
        } catch (e) {
            errors.push("Clusters could not be loaded: " + e.message)
        }

        let jobsWithSchedules = undefined
        try {
            const knownJobs = await call(jobs)()
            jobsWithSchedules = []
            for (const job of knownJobs) {
                let knownSchedules = await call(schedules)(job)
                let lastRun: string | undefined
                let nextRun: string | undefined
                if (knownSchedules !== undefined) {
                    if (knownSchedules.lastRun !== undefined) {
                        lastRun = moment(knownSchedules.lastRun).toISOString()
                    }
                    if (knownSchedules.nextRuns.length > 0) {
                        nextRun = moment(knownSchedules.nextRuns[0]).toISOString()
                    }
                }
                jobsWithSchedules.push({
                    job,
                    lastRun,
                    nextRun,
                    config
                })
            }
        } catch (e) {
            errors.push("Jobs could not be loaded: " + e.message)
        }

        let knownImages = undefined
        try {
            knownImages = await call(list)()
        } catch (e) {
            errors.push("Images could not be loaded: " + e.message)
        }

        return {
            title: "Spacegun ∞ Dashboard",
            clusters: clustersWithNamespaces,
            jobs: jobsWithSchedules,
            images: knownImages,
            config,
            version: process.env.VERSION,
            errors
        }
    }

    @Resource({ path: "/pods/:cluster" })
    public async pods(params: { cluster: string }): Promise<object> {
        const cluster = params.cluster
        const knownNamespaces = await call(namespaces)({ cluster })
        const namespacesWithPods = []

        for (const namespace of knownNamespaces) {
            const knownPods = await call(pods)({ cluster, namespace })

            namespacesWithPods.push({
                name: namespace,
                pods: knownPods
            })
        }

        return {
            title: "Spacegun ∞ Pods ∞ " + cluster,
            name: params.cluster,
            namespaces: namespacesWithPods
        }
    }
    @Resource({ path: "/images/:image" })
    public async images(params: { image: string }): Promise<object> {
        const image = params.image
        const knownTags = await call(tags)({ name: image })
        const versionsWithDates = knownTags.map(tag => ({
            url: "tbd...",
            name: image,
            tag: tag,
        }))

        return {
            title: "Spacegun ∞ Images ∞ " + image,
            name: image,
            versions: versionsWithDates
        }
    }
}

