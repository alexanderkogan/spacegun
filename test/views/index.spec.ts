import { Module } from "../../src/views/index"

const m = new Module()

const mockDispatched = jest.fn()
const mockDispatchFn = jest.fn()
jest.mock("../../src/dispatcher/index", () => ({
    get: (moduleName: string, procedureName: string) => {
        mockDispatched(moduleName, procedureName)
        return mockDispatchFn
    },
    call: (request: any) => {
        mockDispatched(request.module, request.procedure)
        return mockDispatchFn
    },
    add: () => { },
    path: () => ""
}))

describe("view", () => {

    describe("index", () => {

        it("reports errors", async () => {
            mockDispatchFn
                .mockRejectedValueOnce(new Error("error1"))
                .mockRejectedValueOnce(new Error("error2"))
                .mockRejectedValueOnce(new Error("error3"))

            const result = await m.index() as { errors: string[] }

            expect(result.errors).toMatchSnapshot()
        })

        it("fetches namespaces and clusters", async () => {
            mockDispatchFn
                .mockResolvedValueOnce(["cluster1", "cluster2"])
                .mockResolvedValueOnce(["namespace1", "namespace2"])
                .mockResolvedValueOnce(["namespace3"])

            const result = await m.index() as { clusters: object[] }

            expect(result.clusters).toEqual([
                { name: "cluster1", namespaces: ["namespace1", "namespace2"] },
                { name: "cluster2", namespaces: ["namespace3"] }
            ])
        })

        it("fetches pipelines", async () => {
            mockDispatchFn
                .mockResolvedValueOnce([])
                .mockResolvedValueOnce([
                    { name: "pipeline1", cluster: "cluster1", cron: "cron1", start: "start1", steps: [] },
                    { name: "pipeline2", cluster: "cluster2", start: "start1", steps: [] }
                ])
                .mockResolvedValueOnce({ name: "pipeline1", lastRun: 42, nextRuns: [1, 2, 3], isStarted: true, isRunning: true })
                .mockResolvedValueOnce({ name: "pipeline2", isStarted: false, nextRuns: [], isRunning: true })

            const result = await m.index() as { jobs: object[] }

            expect(result.jobs).toEqual([
                {
                    pipeline: { cluster: "cluster1", cron: "cron1", name: "pipeline1", start: "start1", steps: [] },
                    lastRun: "1970-01-01T00:00:00.042Z", nextRun: "1970-01-01T00:00:00.001Z"
                }, {
                    pipeline: { cluster: "cluster2", name: "pipeline2", start: "start1", steps: [] }
                }
            ])
        })

        it("fetches images", async () => {
            mockDispatchFn
                .mockResolvedValueOnce([])
                .mockResolvedValueOnce([])
                .mockResolvedValueOnce(["image1", "image2"])

            const result = await m.index() as { images: object[] }

            expect(result.images).toEqual(["image1", "image2"])
        })
    })

    describe("pods", () => {

        it("fetches namespaces and pods", async () => {
            mockDispatchFn
                .mockResolvedValueOnce(["namespace1", "namespace2"])
                .mockResolvedValueOnce([
                    { justSome: "pod" }
                ])

            const result = await m.pods({ cluster: "cluster1" }) as { namespaces: object[] }

            expect(result.namespaces).toEqual([
                { name: "namespace1", pods: [{ "justSome": "pod" }] },
                { name: "namespace2", pods: undefined }
            ])
        })
    })

    describe("images", () => {

        it("fetches images", async () => {
            mockDispatchFn
                .mockResolvedValueOnce(["tag1", "tag2", "tag3"])

            const result = await m.images({ image: "image1" }) as { images: object[] }

            expect(result.images).toEqual([
                { name: "image1", tag: "tag1" },
                { name: "image1", tag: "tag2" },
                { name: "image1", tag: "tag3" }
            ])
        })

        describe("focused image", () => {

            it("fetches focused image", async () => {
                mockDispatchFn
                    .mockResolvedValueOnce(["tag1", "tag2"])
                    .mockResolvedValueOnce({ focused: "image" })

                const result = await m.images({ image: "image1", tag: "tag1" }) as { focusedImage: object }

                expect(result.focusedImage).toEqual({ focused: "image" })
            })

            it("fetches only if the tag is available", async () => {
                mockDispatchFn
                    .mockResolvedValueOnce(["tag1", "tag2"])

                const result = await m.images({ image: "image1", tag: "unkownTag" }) as { focusedImage: object }

                expect(result.focusedImage).toBeUndefined()
            })

            it("defaults to latest", async () => {
                mockDispatchFn
                    .mockResolvedValueOnce(["latest"])
                    .mockResolvedValueOnce({ focused: "image" })

                const result = await m.images({ image: "image1" }) as { focusedImage: object }

                expect(result.focusedImage).toEqual({ focused: "image" })
            })
        })
    })
})
