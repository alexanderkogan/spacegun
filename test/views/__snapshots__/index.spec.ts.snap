// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view index reports errors 1`] = `
Array [
  "Clusters could not be loaded: error1",
  "Jobs could not be loaded: error2",
  "Images could not be loaded: error3",
]
`;
